File: python/obfuscator.py
#!/usr/bin/env python3
import argparse
import base64
import sys
import os
import getpass
import hashlib
import secrets


def xor_bytes(data: bytes, key_bytes: bytes) -> bytes:
    return bytes([b ^ key_bytes[i % len(key_bytes)] for i, b in enumerate(data)])


def obfuscate(input_path: str, output_path: str, password: str):
    with open(input_path, 'rb') as f:
        src = f.read()

    # Derive a symmetric key from the password using PBKDF2
    salt = secrets.token_bytes(16)
    iterations = 200_000
    key_bytes = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, iterations, dklen=32)

    enc = xor_bytes(src, key_bytes)
    b64 = base64.b64encode(enc).decode('ascii')

    # Emit stub that contains the salt and iterations (so the runtime can derive the key)
    stub = f'''# Obfuscated file (generated)
import base64, sys, os, getpass, hashlib
_b64 = """{b64}"""
_salt_hex = "{salt.hex()}"
_iters = {iterations}

_key = os.environ.get('OBF_KEY')
if not _key:
    if len(sys.argv) > 1:
        _key = sys.argv[1]
    else:
        try:
            _key = getpass.getpass('Password: ')
        except Exception:
            print('Missing key: set OBF_KEY, pass as first arg, or provide via prompt')
            sys.exit(1)

def _xor(data, key_bytes):
    return bytes([b ^ key_bytes[i % len(key_bytes)] for i, b in enumerate(data)])

salt = bytes.fromhex(_salt_hex)
key_bytes = hashlib.pbkdf2_hmac('sha256', _key.encode(), salt, _iters, dklen=32)
code = _xor(base64.b64decode(_b64), key_bytes)
exec(compile(code, '<obfuscated>', 'exec'))
'''

    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(stub)
    print('Wrote obfuscated file:', output_path)


def main():
    p = argparse.ArgumentParser(description='PBKDF2-based obfuscator (XOR keystream)')
    p.add_argument('input')
    p.add_argument('output')
    p.add_argument('--key', required=False, help='Secret password used to derive the key')
    args = p.parse_args()

    password = args.key
    if not password:
        try:
            password = getpass.getpass('Password to derive key: ')
        except Exception:
            print('Provide --key or run interactively')
            sys.exit(2)

    obfuscate(args.input, args.output, password)


if __name__ == '__main__':
    main()

File: go/obfuscator.go
package main

import (
    "crypto/hmac"
    "crypto/rand"
    "crypto/sha256"
    "encoding/base64"
    "encoding/hex"
    "flag"
    "fmt"
    "io"
    "io/ioutil"
    "os"
    "strconv"
)

func pbkdf2(password, salt []byte, iter, keyLen int) []byte {
    hLen := 32 // sha256
    blocks := (keyLen + hLen - 1) / hLen
    dk := make([]byte, 0, blocks*hLen)
    for b := 1; b <= blocks; b++ {
        // U_1 = HMAC(password, salt || INT(b))
        mac := hmac.New(sha256.New, password)
        mac.Write(salt)
        mac.Write([]byte{byte(b >> 24), byte(b >> 16), byte(b >> 8), byte(b)})
        u := mac.Sum(nil)
        t := make([]byte, len(u))
        copy(t, u)
        for i := 1; i < iter; i++ {
            mac = hmac.New(sha256.New, password)
            mac.Write(u)
            u = mac.Sum(nil)
            for j := 0; j < len(t); j++ {
                t[j] ^= u[j]
            }
        }
        dk = append(dk, t...)
    }
    return dk[:keyLen]
}

func xor(data, key []byte) []byte {
    out := make([]byte, len(data))
    for i := range data {
        out[i] = data[i] ^ key[i%len(key)]
    }
    return out
}

func main() {
    in := flag.String("in", "", "input Go file")
    out := flag.String("out", "", "output wrapper file")
    key := flag.String("key", "", "password (optional)")
    flag.Parse()
    if *in == "" || *out == "" {
        fmt.Println("Usage: obfuscator -in input.go -out wrapper.go [-key password]")
        os.Exit(2)
    }

    data, err := ioutil.ReadFile(*in)
    if err != nil {
        fmt.Println("read error:", err)
        os.Exit(1)
    }

    password := *key
    if password == "" {
        fmt.Print("Password: ")
        var b = make([]byte, 128)
        n, _ := os.Stdin.Read(b)
        password = string(b[:n])
        password = string([]byte(password))
        password = password
    }

    salt := make([]byte, 16)
    if _, err := io.ReadFull(rand.Reader, salt); err != nil {
        fmt.Println("salt error:", err)
        os.Exit(1)
    }
    iters := 200000
    keyBytes := pbkdf2([]byte(password), salt, iters, 32)
    enc := xor(data, keyBytes)
    b64 := base64.StdEncoding.EncodeToString(enc)

    // wrapper that decodes and runs the code via `go run`
    wrapper := fmt.Sprintf(`package main

import (
    "crypto/hmac"
    "crypto/sha256"
    "encoding/base64"
    "encoding/hex"
    "fmt"
    "io/ioutil"
    "os"
    "os/exec"
)

func pbkdf2(password, salt []byte, iter, keyLen int) []byte {
    hLen := 32
    blocks := (keyLen + hLen - 1) / hLen
    dk := make([]byte, 0, blocks*hLen)
    for b := 1; b <= blocks; b++ {
        mac := hmac.New(sha256.New, password)
        mac.Write(salt)
        mac.Write([]byte{byte(b >> 24), byte(b >> 16), byte(b >> 8), byte(b)})
        u := mac.Sum(nil)
        t := make([]byte, len(u))
        copy(t, u)
        for i := 1; i < iter; i++ {
            mac = hmac.New(sha256.New, password)
            mac.Write(u)
            u = mac.Sum(nil)
            for j := 0; j < len(t); j++ {
                t[j] ^= u[j]
            }
        }
        dk = append(dk, t...)
    }
    return dk[:keyLen]
}

func xor(data, key []byte) []byte {
    out := make([]byte, len(data))
    for i := range data {
        out[i] = data[i] ^ key[i%len(key)]
    }
    return out
}

func main() {
    b64 := `%s`
    saltHex := "%s"
    iters := %d

    pass := os.Getenv("OBF_KEY")
    if pass == "" {
        if len(os.Args) > 1 {
            pass = os.Args[1]
        } else {
            fmt.Print("Password: ")
            var input []byte
            fmt.Scanln(&input)
            pass = string(input)
        }
    }
    salt, _ := hex.DecodeString(saltHex)
    key := pbkdf2([]byte(pass), salt, iters, 32)
    raw, _ := base64.StdEncoding.DecodeString(b64)
    decoded := xor(raw, key)
    tmp := "_decoded.go"
    ioutil.WriteFile(tmp, decoded, 0644)
    cmd := exec.Command("go", "run", tmp)
    cmd.Stdout = os.Stdout
    cmd.Stderr = os.Stderr
    cmd.Stdin = os.Stdin
    if err := cmd.Run(); err != nil {
        fmt.Fprintln(os.Stderr, "run error:", err)
        os.Exit(1)
    }
}
`, b64, hex.EncodeToString(salt), iters)

    if err := ioutil.WriteFile(*out, []byte(wrapper), 0644); err != nil {
        fmt.Println("write error:", err)
        os.Exit(1)
    }
    fmt.Println("Wrote Go wrapper:", *out)
}
